lib/puppet/indirector/node/rest.rb

```diff
@@ -4,4 +4,14 @@
 class Puppet::Node::Rest < Puppet::Indirector::REST
   desc "Get a node via REST. Puppet agent uses this to allow the puppet master
     to override its environment."
+
+  def find(request)
+    super(request) do |response|
+      server_version = response['X-Puppet-Version']
+      if SemanticPuppet::Version.parse(server_version).major < SemanticPuppet::Version.parse(Puppet::PUPPETVERSION).major
+        Puppet.warning "Server version #{server_version} is unsupported, downgrading to PSON"
+        Puppet[:preferred_serialization_format] = 'pson'
```

Comment:
The node and report rest implementations are slightly different, because in the node case, the agent will send `Accept: application/json, text/pson, ...` and it's likely that a 4.x master will successfully downgrade the response to `text/pson`, and we just need to use JSON for any *future* request.

The report case is different, because if the client `PUT`s the report in json, then an older master will return 500, and the current request will have failed.

That said, I didn't like how the earlier version of this PR modified global state as a result of calling `Node.indirection.find`. I modified both cases to raise exceptions, which I think is cleaner, since the caller can decide what to do.

That said, I'm wondering if we should even attempt to downgrade, or just fail all rest requests when talking to an older master?